job-runner/chain/chain.go

```diff
@@ -97,103 +135,197 @@ func (c *chain) NextJobs(jobId string) proto.Jobs {
 }
 
 // JobIsReady returns whether or not a job is ready to run. A job is considered
-// ready to run if all of its previous jobs are complete. If any previous jobs
-// are not complete, the job is not ready to run.
+// ready to run if it is Pending or Stopped with some retry attempts remaining,
+// and all of its previous jobs are complete. If any previous jobs are not
+// complete, the job is not ready to run.
 func (c *chain) JobIsReady(jobId string) bool {
 	c.jcMux.RLock()
 	defer c.jcMux.RUnlock()
-	for _, job := range c.previousJobs(jobId) {
-		if job.State != proto.STATE_COMPLETE {
-			return false
+	return c.jobIsReady(jobId)
+}
+
+// RunnableJobs returns a list of all jobs that are ready to run. A job is
+// ready to run if all of its previous jobs are complete and it is Pending
+// or Stopped with some retries still remaining.
+func (c *chain) RunnableJobs() proto.Jobs {
```

Comment:
Why does this not return `[]proto.Job`?

Reply:
For testing - proto.Jobs lets you sort the list of jobs to compare it to an expected list. Other than that, proto.Jobs is the same as []proto.Job.